src/server/enterprise/server/api_server.go

```diff
@@ -332,3 +346,237 @@ func (a *apiServer) Deactivate(ctx context.Context, req *ec.DeactivateRequest) (
 
 	return &ec.DeactivateResponse{}, nil
 }
+
+// Pause sets the cluster to paused mode, restarting all pachds in a paused
+// status.  If they are already paused, it is a no-op.
+func (a *apiServer) Pause(ctx context.Context, req *ec.PauseRequest) (resp *ec.PauseResponse, retErr error) {
+	if a.env.Mode == SidecarMode || a.env.Mode == EnterpriseMode {
+		return nil, errors.Errorf("cannot pause a sidecar or enterprise server")
+	}
+	if err := a.rollPachd(ctx, true); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+
+	return &ec.PauseResponse{}, nil
+}
+
+const updatedAtFieldName = "pachyderm.com/updatedAt"
+
+// rollPachd changes pachds from paused to unpaused, or unpaused to paused.  It
+// does this by creating a ConfigMap with a MODE value, then updating the pachd
+// deployment spec’s template’s annotations with a new value, which causes
+// Kubernetes to terminate existing pods and start new ones.
+//
+// Since the ConfigMap is not managed by Helm, it will persist through Helm
+// operations: a cluster which is upgraded by Helm in a paused state will thus
+// remain paused.
+//
+// If the ConfigMap is already in the desired state, no changes are made; this
+// ensures that PauseStatus can do its checking appropriately.
+//
+// There is a special case in main to handle the case where there is no
+// ConfigMap and the '$(MODE)' reference is verbatim.
+func (a *apiServer) rollPachd(ctx context.Context, paused bool) error {
+	var (
+		kc        *kubernetes.Clientset = a.env.GetKubeClient()
+		namespace                       = a.env.Namespace
+	)
+	cc := kc.CoreV1().ConfigMaps(namespace)
+	c, err := cc.Get(ctx, "pachd-config", metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		// if the ConfigMap is not found, then the cluster is by definition unpaused
+		if !paused {
+			return nil
+		}
+		// Since pachd-config is not managed by Helm, it may not exist.
+		c = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pachd-config",
+				Namespace: namespace,
+			},
+			Data: make(map[string]string),
+		}
+		c.Data["MODE"] = "paused"
+		c.Data["previous-mode"] = "full"
```

Comment:
I don't think we know that we're in "full" mode at this point

Reply:
I believe so. This is within the `if k8serrors.IsNotFound(err)` block, which means that there is no ConfigMap. `rollPachd` is only called for full or paused mode, and if there is no ConfigMap then the cluster has to be in full mode.